main/core.js

```diff
@@ -0,0 +1,174 @@
+    env: {
+      FIL_WALLET_ADDRESS: wallet.getAddress()
+    }
+  })
+
+  /** @type {string | null} */
+  let coreExitReason = null
+
+  coreChildProcess.on('close', code => {
+    console.log(`Core closed all stdio with code ${code ?? '<no code>'}`)
+
+    ;(async () => {
+      const log = await getLog()
+      Sentry.captureException('Core exited', scope => {
+        // Sentry UI can't show the full 100 lines
+        scope.setExtra('logs', log.split('\n').slice(-10).join('\n'))
+        scope.setExtra('reason', coreExitReason)
+        return scope
+      })
+    })()
+  })
+
+  coreChildProcess.on('exit', (code, signal) => {
+    const reason = signal ? `via signal ${signal}` : `with code: ${code}`
+    const msg = `Core exited ${reason}`
+    console.log(msg)
+    coreExitReason = signal || code ? reason : null
+  })
+}
+
+function startEventsProcess (/** @type {Context} */ ctx) {
+  console.log('Starting Core Events...')
+  const events = execa(corePath, ['events'])
+  assert(events.stdout, 'Events child process has no stdout')
+  events.stdout
+    .pipe(JSONStream.parse(true))
+    .on('data', (/** @type {CoreEvent} */ event) => {
+      switch (event.type) {
+        case 'jobs-completed': {
+          ctx.setTotalJobsCompleted(event.total)
+          break
+        }
+        case 'activity:info': {
+          ctx.recordActivity({
+            source: event.module,
+            type: 'info',
+            message: event.message,
+            date: new Date(event.date).getTime()
+          })
+          if (event.message.includes('Saturn Node is online')) {
+            online = true
+          }
+          break
+        }
+        case 'activity:error': {
+          ctx.recordActivity({
+            source: event.module,
+            type: 'error',
+            message: event.message,
+            date: new Date(event.date).getTime()
+          })
+          if (
+            event.message === 'Saturn Node started.' ||
+            event.message.includes('was able to connect') ||
+            event.message.includes('will try to connect')
+          ) {
+            online = false
+          }
+        }
+      }
+    })
+  app.on('before-quit', () => {
+    events.kill()
+  })
+  events.on('exit', (code, signal) => {
+    if (code !== 0) {
+      console.log(`Events process exited with code=${code} signal=${signal}`)
+      setTimeout(() => startEventsProcess(ctx), 1000)
+    }
+  })
+}
+
+function isOnline () {
+  return online
+}
+
+/**
+ * @returns {Promise<Activity[]>}
+ */
+async function getActivity () {
+  const { stdout: activity } = await execa(
+    corePath,
+    ['activity', '--json']
+  )
+  return JSON.parse(activity)
+}
+
+async function getLog () {
+  const { stdout: log } = await execa(
+    corePath,
+    ['logs'],
+    { encoding: 'utf8' }
+  )
+  return log
+}
+
+async function getMetrics () {
+  const { stdout: json } = await execa(
+    corePath,
+    ['metrics']
+  )
+  return JSON.parse(json)
+}
```

Comment:
As discussed in a call, the current architecture decouples `core` from `desktop` on the process level, so that eventually `core` can be rewritten in a different language (go or rust probably).

Reply:
Actually, after my last night walk, I'm now convinced that this is the way, because it will make the integration so much easier. And the APIs are mostly already there, just need to be documented:

```js
// get metrics
const metrics = await metrics.getLatestMetrics()

// stream metrics
for await (const obj of await metrics.followMetrics()) {
  // ...
}

// etc
```

Reply:
<img width="632" alt="Screenshot 2023-04-20 at 11 27 46" src="https://user-images.githubusercontent.com/10247/233322963-8b400ac7-d807-4c63-845f-237a4f5eea2d.png">

@bajtos do you know what the way to go is here?

Reply:
Looks like we could

1. Transpile to CommonJS
2. Change to CommonJS
3. Use dynamic imports

Reply:
No need to transpile. You need to load Station Core via dynamic import from our CommonJS files.

```js
const coreLog = await import('@filecoin-station/core/lib/log.js');
```

I am not sure if Node.js' ESM loader supports loading internal package files directly, as I shown in my code snippet. If it does not, then we can:
1. Modify `package.json` in Station Core to tell Node.js which files can be loaded
2. Rework Station Core to export all public APIs from the main file, so that we don't need to import from internal files.

I prefer the second option.

```js
const core = await import('@filecoin-station/core')
const log = core.log;
const metrics = core.metrics;
// etc.
```